spec.bs

```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If |urls| is empty or exceeds the maximum allowed length, return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |urlList| be an empty {{list}}.
+    1. [=map/iterate|For each=] |urlWithMetadata| in |urls|:
+        1. If |urlWithMetadata| has no field "`url`", return a [=promise rejected=] with a {{TypeError}}.
+        1. Otherwise, let |urlString| be |urlWithMetadata|["`url`"].
+        1. Let |serializedUrl| be the result of running [=get the canonical URL string if valid=] with |urlString|.
+        1. If |serializedUrl| is undefined, return a [=promise rejected=] with a {{TypeError}}.
+        1. Otherwise, [=list/append=] |serializedUrl| to |urlList|.
+        1. If |urlWithMetadata| has field "`reportingMetadata`":
+            1. Let |reportingMetadata| be |urlWithMetadata|["`reportingMetadata`"].
+            1. If the result of running [=validate reporting metadata=]  with |reportingMetadata| is false, reject |resultPromise| with a {{TypeError}} and abort these steps.
+    1. Let |fencedFrameConfigStruct| be a [=fenced frame config=] with an invalid "urn uuid". <span class=todo>Add correct struct name as well as linking when Fenced Frame API updates their draft spec to include it.</span>
+    1. Let |environment| be |window|'s [=relevant settings object=].
+    1. If the result of running [=determine whether shared storage is allowed by context=] given |environment| and [=this=]'s [=SharedStorageWorklet/worklet origin=] is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. If the result of running [=determine whether shared storage is allowed by enrollment and user preference=] given |environment| and [=this=]'s [=SharedStorageWorklet/worklet origin=] is false:
+        1. If [=this=]'s [=SharedStorageWorklet/is cross-origin worklet=] is false, return a [=promise rejected=] with a {{TypeError}}.
+    1. Else, set |fencedFrameConfigStruct| to a [=fenced frame config=] with a pending mapped "url uuid". <span class=todo>Add correct struct name as well as linking when Fenced Frame API updates their draft spec to include it.</span>
+
+        Issue: The [=fenced frame config=] and the following "obtain a {{FencedFrameConfig}} from a fenced frame config struct" algorithm have not yet been added to the draft [=Fenced Frame=] specification. Some form of them will be added, although their names are subject to bikeshedding.
+
+    1. If |options|["`resolveToConfig`"] is true, resolve |resultPromise| with the result of running "obtain a {{FencedFrameConfig}} from a fenced frame config struct" with |fencedFrameConfigStruct|. <span class=todo>Add correct struct and algorithms names as well as linking when Fenced Frame API updates their draft spec to include it.</span>
+    1. Othewise, resolve |resultPromise| to |fencedFrameConfigStruct|'s "urn uuid". <span class=todo>Add correct struct name and urn:uuid name as well as linking when Fenced Frame API updates their draft spec to include it.</span>
```

Comment:
Should exist by now.

Reply:
Hmm.. the internal struct type ([FencedFrameConfig](https://wicg.github.io/fenced-frame/#fenced-frame-config); the one we are using here) still _does not_ contain any urn uuid type. Only the JavaScript type ([FencedFrameConfig](https://wicg.github.io/fenced-frame/#fencedframeconfig)) has it. Anyway, I feel it could be done separately as there are a lot of other places referring to FF/TODO, and the change seems orthogonal to this patch.

Reply:
Well we should at least check with people active over on that specification to see if there are plans to address that TODO. This TODO (and others?) make sound like the FF spec is planning on adding a URN UUID member to the struct, but to my knowledge they are not. We should make the need for this more vocally. @gtanzer, what do you think? Can we help resolve this?

Reply:
I've never really understood in this:

"One major input to the fencedframe element is the FencedFrameConfig interface, which maps to an internal fenced frame config struct."

what exactly "maps to an internal fenced frame config struct" means for an interface in spec. It really maps to a pair of a urn uuid and a fenced frame config struct anyway.

Regardless, there are way more TODOs than are necessary. All of the stuff about pending configs / generating urns uses the interface in fenced rame config mapping.
https://wicg.github.io/fenced-frame/#fenced-frame-config-mapping

e.g. here in the PA spec:
https://wicg.github.io/turtledove/#:~:text=Let%20fencedFrameConfig%20be%20the%20result%20of%20filling%20in%20a%20pending%20fenced%20frame%20config%20with%20pendingConfig%2C%20auctionConfig%2C%20winnerInfo%2C%20and%20auctionReportInfo.

which is a bit sloppy with the urn part.

I think the issue is that originally we intended not to use urn:uuids in spec, since they are only necessary in implementation as a browser/renderer thing. In spec you can just use references to the same underlying config object.

But urn iframes complicate everything, and then I think a bunch of stuff is spec'd in terms of lookups into the config mapping using the urn...